src/libAtomVM/ets_hashtable.c

```diff
@@ -98,15 +138,17 @@ EtsHashtableErrorCode ets_hashtable_insert(struct EtsHashTable *hash_table, term
         while (1) {
             if (term_compare(key, node->key, TermCompareExact, global) == TermEquals) {
                 if (opts & EtsHashtableAllowOverwrite) {
-                    node->entry = entry;
+                    node->entry = new_node->entry;
                     memory_destroy_heap(node->heap, global);
-                    node->heap = heap;
+                    node->heap = new_node->heap;
+                    free(new_node);
                     return EtsHashtableOk;
                 } else {
+                    memory_destroy_heap(new_node->heap, global);
+                    free(new_node);
```

Comment:
I don't like this, we shouldn't take ownership of the node here.

Reply:
Did I miss something?

Reply:
Not sure, I don't remember talking offline about this. The point is: if we don't allow overwrite then we should only return `EtsHashtableFailure` instead of doing some weird side-effects. We can take ownership of the node if insert is successful.

Reply:
Still not sure if I'm getting this, this is one of the situation where sharing a snippet or some schema might be useful.

Reply:
So what I'm saying is when you have this kind of usage:
```
struct HNode *new_node = ets_hashtable_new_node(...);
EtsHashtableErrorCode result = ets_hashtable_insert(table, new_node);
if(result == EtsHashtableFailure) {
    // handle error, retry, try something different with the new node
}
```

You know that insert wasn't successful for some reason.

My intuition is that `new_node` was unchanged and I can do whatever I want since it failed. In the code, however, we clear the new node's heap so it's unusable regardless of what code is returned.

For success case I more easily accept that I gave up my control over node and it's invalid after the call.